Convert to admission webhook-recognized versions

liggitt opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Implements converting objects to webhook-understood versions, to allow webhooks to interoperate with a variety of Kubernetes releases, even ones supporting new versions of resources the webhook intercepts, but which did not exist when the webhook was written.

**Which issue(s) this PR fixes**:
xref https://github.com/kubernetes/enhancements/issues/492
xref https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#convert-to-webhook-requested-version

**Special notes for your reviewer**:

Best reviewed a commit at a time, changes follow a fairly reasonable sequence

**Does this PR introduce a user-facing change?**:
```release-note
Admission webhooks can now register for a single version of a resource (for example, `apps/v1 deployments`) and be called when any other version of that resource is modified (for example `extensions/v1beta1 deployments`). This allows new versions of a resource to be handled by admission webhooks without needing to update every webhook to understand the new version. See the API documentation for the `matchPolicy: Equivalent` option in MutatingWebhookConfiguration and ValidatingWebhookConfiguration types.
```

Follow-ups:
- [x] clarify requestKind, requestResource, requestSubResource are always provided in v1 AdmissionReview
- [x] document possibility of CRD conversion being asked to convert pre-schema validation

/sig api-machinery
/cc @sttts @jpbetz @caesarxuchao

liggitt committed:
make ObjectInterfaces impl generic

liggitt committed:
Add GetResourceMapper to admission ObjectInterfaces

liggitt committed:
Move object conversion to webhook dispatch point

convert versionedattrs as needed

Allow per-webhook kind/version

liggitt committed:
Consider equivalent resources when calling webhook

liggitt committed:
AdmissionReview API changes

liggitt committed:
AdmissionRegistration API changes: MatchPolicy

liggitt committed:
generated: AdmissionReview changes

liggitt committed:
generated: AdmissionRegistration changes

liggitt committed:
Skip namespace selector evaluation for 'select all' selectors

liggitt commented:
/priority important-soon

liggitt committed:
Map replicationcontrollers/scale to extensions/v1beta1 replicationcontrollers/scale

liggitt committed:
Add webhook admission conversion test

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/util.go:
```diff
@@ -18,11 +18,26 @@ package admission
 
-type SchemeBasedObjectInterfaces struct {
-	Scheme *runtime.Scheme
+type DefaultObjectInterfaces struct {
```
"default" is an overloaded term. RuntimeObjectInterfaces?

sttts commented on pkg/apis/admission/types.go:
```diff
@@ -56,6 +56,27 @@ type AdmissionRequest struct {
+	// If this differs from the value in "kind", an equivalent match and conversion was performed.
+	// See documentation for the "matchPolicy" field in the webhook configuration type.
+	// +optional
+	RequestKind *metav1.GroupVersionKind
```
it's unclear what this is when there was not conversion happening? Will it be nil? Or the equal value? Same below.

liggitt replied:
we will always populate it, but old apiservers will not (obviously), so someone writing a webhook using our go types to work against pre-1.15 apiservers can not assume they'll get a value here. we can strengthen the spec in v1 admissionreview

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -76,6 +76,8 @@ type ObjectInterfaces interface {
 	// GetObjectConvertor is the ObjectConvertor appropriate for the requested object.
 	GetObjectConvertor() runtime.ObjectConvertor
+	// GetResourceMapper is the ResourceMapper appropriate for finding equivalent resources and expected kind for the requested object.
+	GetResourceMapper() runtime.ResourceMapper
```
The other interface are retrieve-only, this one allows mutation. It makes it hard to reason about it when the ObjectInterfaces are passed around (like to admission). I would prefer to pass a read-only variant to admission instead.

liggitt replied:
split into readonly and writeable interfaces

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -157,6 +157,10 @@ type GenericAPIServer struct {
 	Authorizer authorizer.Authorizer
 
+	// ResourceMapper provides information about resources equivalent to a given resource,
+	// and the kind associated with a given resource
```
... This is updated with each API that is installed.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -188,6 +189,10 @@ type Config struct {
+
+	// ResourceMapper provides information about equivalent resources,
+	// and the kind associated with a given resource
+	ResourceMapper runtime.ResourceMapper
```
also here it is unclear whether this provides or receives that information. Having two interfaces for read and write would make that much clearer.

ResourceMappingRecorder vs. ResourceMapper

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:
```diff
@@ -532,6 +532,8 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		UnsafeConvertor: a.group.UnsafeConvertor,
 		Authorizer:      a.group.Authorizer,
 
```
why empty line? Is that different?

liggitt replied:
just to avoid big diffs on aligned fields

sttts commented on staging/src/k8s.io/apimachinery/pkg/runtime/mapper.go:
```diff
@@ -0,0 +1,87 @@
+func (r *resourceMapper) KindFor(resource schema.GroupVersionResource, subresource string) schema.GroupVersionKind {
+	return r.kinds[resource][subresource]
+}
+func (r *resourceMapper) SetKindFor(resource schema.GroupVersionResource, subresource string, kind schema.GroupVersionKind) {
```
Document that this is not Go routine safe.

liggitt replied:
made safe

sttts commented on staging/src/k8s.io/apimachinery/pkg/runtime/mapper.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+// NewResourceMapperWithKeyFunc creates a resource mapper with a custom key function.
+// The provided function is called to determine the key for each group resource.
```
so it is actually an identity func.

liggitt replied:
changed name to identity func and updated doc

sttts commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -210,6 +210,18 @@ type ObjectCreater interface {
 	New(kind schema.GroupVersionKind) (out Object, err error)
 }
 
+// ResourceMapper provides information about equivalent resources,
```
this needs more explanation: resources are GVRs. How can GVRs be equivalent if they are not equal?

liggitt replied:
updated docs

sttts commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -210,6 +210,18 @@ type ObjectCreater interface {
+type ResourceMapper interface {
+	// ResourcesFor returns a list of resources equivalent to resource.
+	// If subresource is specified, only equivalent resources which have the same subresource are returned.
+	ResourcesFor(resource schema.GroupVersionResource, subresource string) []schema.GroupVersionResource
```
`EquivalentResourcesFor`. `ResourcesFor` is underspecified.

liggitt replied:
done

sttts commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -210,6 +210,18 @@ type ObjectCreater interface {
 }
 
+// ResourceMapper provides information about equivalent resources,
+// and the kind associated with a given resource
```
missing the role of subresources. Do you identify gvr+subresource with other gvr+subresource values?

liggitt replied:
documented meaning of subresource in each function

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/conversion.go:
```diff
@@ -41,3 +41,72 @@ func ConvertToGVK(obj runtime.Object, gvk schema.GroupVersionKind, o admission.O
+// * attr.VersionedObject is used as the source for the new object if Dirty=true (and is round-tripped through attr.Attributes.Object, clearing Dirty in the process)
+// * attr.Attributes.Object is used as the source for the new object if Dirty=false
+// * attr.Attributes.OldObject is used as the source for the old object
+func ConvertVersionedAttributesToGVK(attr *VersionedAttributes, gvk schema.GroupVersionKind, o admission.ObjectInterfaces) error {
```
s/ToGVK//

liggitt replied:
dne

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/conversion.go:
```diff
@@ -41,3 +41,72 @@ func ConvertToGVK(obj runtime.Object, gvk schema.GroupVersionKind, o admission.O
+// * attr.Attributes.Object is used as the source for the new object if Dirty=false
+// * attr.Attributes.OldObject is used as the source for the old object
+func ConvertVersionedAttributesToGVK(attr *VersionedAttributes, gvk schema.GroupVersionKind, o admission.ObjectInterfaces) error {
+	// If we're already the desired kind, we're done
```
nit: we have already...

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -125,21 +125,33 @@ func (a *Webhook) ValidateInitialization() error {
-// ShouldCallHook makes a decision on whether to call the webhook or not by the attribute.
-func (a *Webhook) ShouldCallHook(h *v1beta1.Webhook, attr admission.Attributes) (bool, *apierrors.StatusError) {
-	var matches bool
+// shouldCallHook makes a decision on whether to call the webhook or not by the attribute.
```
nit: by the attributes

sttts replied:
... and returns nil if not.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -154,14 +166,14 @@ func (a *Webhook) Dispatch(attr admission.Attributes, o admission.ObjectInterfac
+	var relevantHooks []*WebhookInvocation
 	for i := range hooks {
-		call, err := a.ShouldCallHook(&hooks[i], attr)
+		call, err := a.shouldCallHook(&hooks[i], attr)
```
s/call/invocation/

liggitt commented:
addressed comments in individual fixup commits (or the delta is usable as well: https://github.com/kubernetes/kubernetes/compare/de217a168c4c4654612c12b8141c21552087cc21..5644f88ed35a25e4ce274e1d011f2aae69428a5a)

sttts commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -177,6 +187,17 @@ type Webhook struct {
+	// via API version v1 and v2, and "rules" only includes "v1", do not send a request to "v2" to the webhook.
+	// - Equivalent: match requests if they modify a resource listed in rules via another API group or version.
+	// For example, if an object can be modified via API version v1 and v2, and "rules" only includes "v1",
+	// a request to "v2" should be converted to "v1" and sent to the webhook.
```
Without colocation in mind this doc really sounds cryptic. "Convertible" would maybe be a better term than the super generic "Equivalent".

If we stay with Equivalent, something like this is much more concrete:

match requests modifying resources listed in rules exactly, or those in different version in the same group, or those which are colocated in storage, i.e. represent the same objects.

jpbetz replied:
Agree. Expanding on "via another API group or version" should help, maybe like @sttts suggests.

I'm not sure if we should introduce the colocated term (which I'm guessing is unfamiliar to most readers and needs to be explained a bit?) or if should try to build on intuition, e.g. "if an object can be modified at either API group version, they are equivalent"...

liggitt replied:
I think "equivalent" is better than "convertible". An apps/v1 Deployment object sent to an `apps/v1/deployments/.../status` *subresource* can be converted to an apps/v1beta1 Deployment, but would not be sent to a webhook registered for `apps/v1beta1/deployments` (not registered for the status subresource)

I'll update the doc to try to clarify this

liggitt replied:
updated to:
```
// matchPolicy defines how the "rules" list is used to match incoming requests.
// Allowed values are "Exact" or "Equivalent".
//
// - Exact: match a request only if it exactly matches a specified rule.
// For example, if deployments can be modified via apps/v1, apps/v1beta1, and extensions/v1beta1,
// but "rules" only included `apiGroups:["apps"], apiVersions:["v1"], resources: ["deployments"]`,
// a request to apps/v1beta1 or extensions/v1beta1 would not be sent to the webhook.
//
// - Equivalent: match a request if modifies a resource listed in rules, even via another API group or version.
// For example, if deployments can be modified via apps/v1, apps/v1beta1, and extensions/v1beta1,
// and "rules" only included `apiGroups:["apps"], apiVersions:["v1"], resources: ["deployments"]`,
// a request to apps/v1beta1 or extensions/v1beta1 would be converted to apps/v1 and sent to the webhook.
//
// Defaults to "Exact"
// +optional
```

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -142,6 +142,34 @@ func (a *Webhook) shouldCallHook(h *v1beta1.Webhook, attr admission.Attributes)
+						Subresource: attr.GetSubresource(),
+						Kind:        o.GetResourceMapper().KindFor(equivalent, attr.GetSubresource()),
+					}
+					break
```
you can also use labels to break the outer loop

sttts commented:
Do we have a test that the ResourceMapper is a complete equivalence matching the shared key spaces in etcd? Thinking about something around the etcd storage test by @enj checking that the resource mapper considers GVRs equivalent if the etcd key prefixes are equal.

sttts commented:
(disclaimer: haven't gone through all tests) are there tests for CRDs?

sttts commented on pkg/kubeapiserver/default_storage_factory_builder.go:
```diff
@@ -111,6 +111,7 @@ func (c *completedStorageFactoryConfig) New() (*serverstorage.DefaultStorageFact
 	storageFactory.AddCohabitatingResources(apps.Resource("daemonsets"), extensions.Resource("daemonsets"))
 	storageFactory.AddCohabitatingResources(apps.Resource("replicasets"), extensions.Resource("replicasets"))
 	storageFactory.AddCohabitatingResources(api.Resource("events"), events.Resource("events"))
+	storageFactory.AddCohabitatingResources(api.Resource("replicationcontrollers"), extensions.Resource("replicationcontrollers")) // to make scale subresources equivalent
```
why wasn't that there before and which influence had that?

liggitt replied:
extensions/v1beta1 replicationcontrollers *only* serve a `scale` subresource, no parent resource, so no storage asked about the prefix for an `extensions/v1beta1` replicationcontrollers resource previously. We now need to know what storage is modified by the scale subresource, and the storage factory is the thing we use to answer that question, so I needed to register this cross-group relationship

liggitt commented:
> (disclaimer: haven't gone through all tests) are there tests for CRDs?

yes, the integration test covers conversion for all types, including colocated ones and crds (that's how I found the "missing" extensions/v1beta1 replicationcontroller line)

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -142,6 +142,34 @@ func (a *Webhook) shouldCallHook(h *v1beta1.Webhook, attr admission.Attributes)
+			// see if the rule matches any of the equivalent resources
+			for _, equivalent := range equivalents {
+				if equivalent == attr.GetResource() {
+					// we already checked this
```
odd optimization. Do you expect many rules with just one resource?

liggitt replied:
we already tried to match attr.GetResource() and failed, no need to check it again

liggitt replied:
> You skip a rule if there is only one equivalent and the previous rule was not matching. Why? Doesn't this mean that you even skip every rule than first rules?

no, we skip checking a particular item in the equivalent list if it is the same as attr.GetResource() (which we already checked above)

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -125,21 +125,62 @@ func (a *Webhook) ValidateInitialization() error {
+		// honor earlier rules first
+		for _, r := range h.Rules {
+			// see if the rule matches any of the equivalent resources
+			for _, equivalent := range equivalents {
```
Do we care about ordering of equivalents? E.g. prefer v1 over v1beta1 if it's available?

liggitt replied:
the returned list is in the order registered

the order registered for CRDs is the order listed in the CRD spec.versions list

the order registered for in-tree resources is the order installed, which is the order we list in discovery (for example, deployments prefers this order: extensions/v1beta1, apps/v1, apps/v1beta1, apps/v1alpha1)

that order seems reasonable to me

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -210,6 +210,23 @@ type ObjectCreater interface {
+	// EquivalentResourcesFor returns a list of resources that address the same underlying data as resource.
+	// If subresource is specified, only equivalent resources which also have the same subresource are included.
+	EquivalentResourcesFor(resource schema.GroupVersionResource, subresource string) []schema.GroupVersionResource
+	// KindFor returns the kind expected by the specified resource[/subresource]
```
Document that an zero value `schema.GroupVersionKind` is returned if the mapper has no kind for the provide resource subresource?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -125,21 +125,62 @@ func (a *Webhook) ValidateInitialization() error {
+						Webhook:     h,
+						Resource:    equivalent,
+						Subresource: attr.GetSubresource(),
+						Kind:        o.GetEquivalentResourceMapper().KindFor(equivalent, attr.GetSubresource()),
```
Looks like `KindFor` should never return return a zero value here if the mapper is working correctly, but if it somehow did should we fail fast with an error?

liggitt replied:
added a check, return with an error, and a test

jpbetz reviewed (commented):
Looks great, a couple minor comments/questions

liggitt commented:
addressed comments, diff visible in https://github.com/kubernetes/kubernetes/compare/5644f88ed35a25e4ce274e1d011f2aae69428a5a..4997519269e93cfffff52c61ac697b90d9056bfb

liggitt commented:
/retest

sttts commented on pkg/apis/admission/types.go:
```diff
@@ -56,6 +56,27 @@ type AdmissionRequest struct {
 	// +optional
 	SubResource string
+
+	// RequestKind is the type of object being manipulated by the the original API request.  For example: Pod
```
v1.Pod

sttts commented on pkg/apis/admission/types.go:
```diff
@@ -56,6 +56,27 @@ type AdmissionRequest struct {
+	// See documentation for the "matchPolicy" field in the webhook configuration type.
+	// +optional
+	RequestKind *metav1.GroupVersionKind
+	// RequestResource is the name of the resource being requested by the the original API request.  This is not the kind.  For example: pods
```
v1.pods

sttts commented on pkg/apis/admission/types.go:
```diff
@@ -56,6 +56,27 @@ type AdmissionRequest struct {
 	SubResource string
+
+	// RequestKind is the type of object being manipulated by the the original API request.  For example: Pod
+	// If this differs from the value in "kind", an equivalent match and conversion was performed.
```
If this is non-nil and differs...

Also below for resource.

sttts commented on pkg/apis/admission/types.go:
```diff
@@ -56,6 +56,27 @@ type AdmissionRequest struct {
+	RequestResource *metav1.GroupVersionResource
+	// RequestSubResource is the name of the subresource being requested by the the original API request.  This is a different resource, scoped to the parent
+	// resource, but it may have a different kind. For instance, /pods has the resource "pods" and the kind "Pod", while
+	// /pods/foo/status has the resource "pods", the sub resource "status", and the kind "Pod" (because status operates on
```
odd to get an example where the kind is the same. Expected the opposite, e.g. for scale after "for instance".

liggitt replied:
yeah, most of this was copied/pasted from the existing SubResource doc. will update both to be better

sttts commented on pkg/apis/admission/types.go:
```diff
@@ -56,6 +56,27 @@ type AdmissionRequest struct {
+	// See documentation for the "matchPolicy" field in the webhook configuration type.
+	// +optional
+	RequestResource *metav1.GroupVersionResource
+	// RequestSubResource is the name of the subresource being requested by the the original API request.  This is a different resource, scoped to the parent
```
nit: misleading. Above resource = GVR. here resource = just the subresource without version and group.

liggitt replied:
updated examples and added "fully-qualified" to the kind and resource field doc

sttts commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -177,6 +187,22 @@ type Webhook struct {
+	// - Equivalent: match a request if modifies a resource listed in rules, even via another API group or version.
+	// For example, if deployments can be modified via apps/v1, apps/v1beta1, and extensions/v1beta1,
+	// and "rules" only included `apiGroups:["apps"], apiVersions:["v1"], resources: ["deployments"]`,
+	// a request to apps/v1beta1 or extensions/v1beta1 would be converted to apps/v1 and sent to the webhook.
```
:+1:

sttts commented on staging/src/k8s.io/apimachinery/pkg/runtime/mapper.go:
```diff
@@ -0,0 +1,97 @@
+	// keyFunc computes a key for the specified resource (this allows honoring colocated resources across API groups).
+	// if null, or if "" is returned, resource.String() is used as the key
+	keyFunc func(resource schema.GroupResource) string
+	// resources maps key -> subresource -> equivalent resources (subresource is not included in the returned resources)
```
I guess empty subresource = main resource?

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/conversion.go:
```diff
@@ -41,3 +41,72 @@ func ConvertToGVK(obj runtime.Object, gvk schema.GroupVersionKind, o admission.O
+	if attr.VersionedObject != nil {
+		// convert the existing versioned object to internal
+		if attr.Dirty {
+			err := o.GetObjectConvertor().Convert(attr.VersionedObject, attr.Attributes.GetObject(), nil)
```
this case feels asymmetric: if `attr.VersionedKind == gvk` we don't write back to `GetObject()´. But if we convert, we do. Why don't we keep converting from versioned object to versioned object, keeping attr dirty?

liggitt replied:
> this case feels asymmetric: if attr.VersionedKind == gvk we don't write back to `GetObject()´. But if we convert, we do. Why don't we keep converting from versioned object to versioned object, keeping attr dirty?

The goal of this method is to provide a VersionedObject in the requested version as cheaply as possible. If we already have the requested version, we no-op.

If we don't have the requested verison, we *have* to convert via `attr.GetObject()` because that's our internal version (we can't convert directly from extension/v1beta1 Deployment to apps/v1 Deployment). As long as we're updating attr.GetObject() to match VersionedObject anyway, we might as well clear the dirty flag to avoid needing to redo the same conversion at the end of the mutating Dispatch() method

I'll update the docs on the fields and this method to try to clarify that

sttts replied:
right. Am too much in the CRD world without internal versions. Nvmd.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go:
```diff
@@ -33,13 +34,29 @@ type Source interface {
+	// Attributes holds the original admission attributes
+	Attributes admission.Attributes
+	// VersionedOldObject holds Attributes.OldObject (if non-nil), converted to VersionedKind
 	VersionedOldObject runtime.Object
```
// must never be mutated

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go:
```diff
@@ -33,13 +34,29 @@ type Source interface {
 	VersionedOldObject runtime.Object
-	VersionedObject    runtime.Object
+	// VersionedObject holds Attributes.Object (if non-nil), converted to VersionedKind
+	VersionedObject runtime.Object
```
// may be mutated, but the consumer of this struct has to update `Dirty`.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go:
```diff
@@ -33,13 +34,29 @@ type Source interface {
+	Dirty bool
+}
+
+// WebhookInvocation describes how to call a webhook, including the resource and subresource the webhook registered for that matched
```
... for that matched?

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/util.go:
```diff
@@ -0,0 +1,29 @@
+	resource schema.GroupVersionResource
+}
+
+func (a *attrWithResourceOverride) GetResource() schema.GroupVersionResource { return a.resource }
```
such a small struct, and such a big and ugly util file. Why not hide it somewhere else at the bottom.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -125,21 +125,68 @@ func (a *Webhook) ValidateInitialization() error {
-func (a *Webhook) ShouldCallHook(h *v1beta1.Webhook, attr admission.Attributes) (bool, *apierrors.StatusError) {
-	var matches bool
+// shouldCallHook makes a decision on whether to call the webhook or not based on the specified attributes,
+// and returns `nil, nil` if the webhook should not be called, or `nil, err` if an error was encountered during evaluation.
```
// nil means not to call the webhook.

sttts commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -210,6 +210,24 @@ type ObjectCreater interface {
+type EquivalentResourceMapper interface {
+	// EquivalentResourcesFor returns a list of resources that address the same underlying data as resource.
+	// If subresource is specified, only equivalent resources which also have the same subresource are included.
+	EquivalentResourcesFor(resource schema.GroupVersionResource, subresource string) []schema.GroupVersionResource
```
resource is part of the returned slice if the subresource in defined for that resource.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -125,21 +125,68 @@ func (a *Webhook) ValidateInitialization() error {
+	if invocation == nil && h.MatchPolicy != nil && *h.MatchPolicy == v1beta1.Equivalent {
+		attrWithOverride := &attrWithResourceOverride{Attributes: attr}
+		equivalents := o.GetEquivalentResourceMapper().EquivalentResourcesFor(attr.GetResource(), attr.GetSubresource())
+		if len(equivalents) == 1 {
```
you can skip this if actually. The for loop below is simple enough and continues anyway.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -55,11 +55,29 @@ func newMutatingDispatcher(p *Plugin) func(cm *webhook.ClientManager) generic.Di
+func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces, relevantHooks []*generic.WebhookInvocation) error {
+	var (
+		versionedAttr *generic.VersionedAttributes
+		err           error
```
why this suspicious global err? It is redeclared at least in line 79. If there is no good reason, make it local.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -417,6 +422,21 @@ func (c *Config) Complete(informers informers.SharedInformerFactory) CompletedCo
+			c.EquivalentResourceRegistry = runtime.NewEquivalentResourceRegistry()
+		} else {
+			c.EquivalentResourceRegistry = runtime.NewEquivalentResourceRegistryWithIdentity(func(groupResource schema.GroupResource) string {
+				// use the storage prefix as the key if possible
```
clever

sttts commented:
Looks really clean now. Left a number of nit comments.

sttts commented:
/lgtm
/hold

Feel free to unhold and retag when the nits are fixed.

sttts commented:
needs a rebase

jpbetz commented:
/lgtm

liggitt commented:
comments addressed in https://github.com/kubernetes/kubernetes/compare/4997519269e93cfffff52c61ac697b90d9056bfb..50f6c2fbe8b8b236fbec581c1de796e752fb2f6f

rebased in https://github.com/kubernetes/kubernetes/compare/50f6c2fbe8b8b236fbec581c1de796e752fb2f6f..8c194ea615459ddab09938528eca3a3a228da8a3

liggitt commented:
tagging per https://github.com/kubernetes/kubernetes/pull/78135#issuecomment-496271005

liggitt commented:
/retest

liggitt commented:
/retest
